File: python/0062.py
# https://leetcode.com/problems/unique-paths/

'''
1. Simply count all possible combinations of (m-1) move down and (n-1) move right.
    The formula is (m-1 + n-1)! / (m-1)!(n-1)!

2. Dynamic Programming

Let dp[i][j] denote the number of unqiue paths to arrive the point (i, j) in the m x n grid.
Since the robot can only move right or down, state equation: dp[i][j] = dp[i][j-1] + dp[i-1][j]

            (i-1, j)
                |
                V
(i, j-1)  ->  (i, j)

with initial condition dp[0, :] = dp[:, 0] = 1

Example:

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[1 for _ in range(n)] for _ in range(m)]
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i][j-1] + dp[i-1][j]
        return dp[-1][-1]

3. Improved version of DP

space complexity: O(n), the idea is we only need to store the state of columns in each row

'''

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [1 for _ in range(n)]
        for _ in range(1, m):
            for j in range(1, n):
                dp[j] += dp[j - 1]
        return dp[-1]

File: cpp/0062.cpp
#include <cassert>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<int> row(n, 1);
        for (int i = 1; i < m; i++)
            for (int j = 1; j < n; j++)
                row[j] += row[j-1];
        return row[n-1];
    }
};

int main() {
    Solution sol;
    assert(sol.uniquePaths(3, 7) == 28);
    assert(sol.uniquePaths(7, 3) == 28);
    assert(sol.uniquePaths(3, 2) == 3);
    assert(sol.uniquePaths(3, 3) == 6);
    return 0;
}
